File: grind-75/typescript/704. Binary Search.ts
const binarySearch = (values: number[], target: number) => {

    let lower = 0;
    let upper = values.length-1;

    while (lower<=upper) {
        const middle = Math.floor(lower + ( upper - lower ) / 2);

        // if the midpoint value exceeds our target it means we need to search in the lower bound -> move upper to below midpoint
        if (values[middle] > target) upper = middle-1;
        // if the midpoint value is less than our target is means we need to search in the upper bound -> move lower to above midpoint
        else if (values[middle] < target) lower = middle+1;
        // otherwise we found our boi
        else return middle;
    }

    return -1;
}

File: grind-75/python/704. Binary Search.py
from typing import List

class Solution:
    def search(self, nums: List[int], target: int) -> int:
        
        left = 0
        right = len(nums)-1

        while left <= right:
            mid = left + (right - left) // 2
            # if the midpoint is greater than our target then we need to search in lower bound ( move right to below midpoint )
            if (nums[mid] > target): right = mid-1
            # if the midpoint is lesser than our target then we need to search in the upper bound ( move the left to above midpoint )
            elif (nums[mid] < target): left = mid+1
            # otherwise we've found our target
            else: return mid

        return -1